### ISSUE TITLE
make lint 'prefer typedefs instead of Function signature defining"

### ISSUE THREAD

**Issue by subzero911** [dart-lang/linter#2733]
It should prefer built-in typedefs: `VoidCallback, ValueSetter<T>, ValueChanged<T>` etc. over plain `void Function (String)` signature.
Many developers even don't know about these typedefs. It makes code look cleaner.

**Comment by bwilkerson** [comment]
This is a very interesting idea, but I have a concern about false positives. I also have a suggestion for how to modify the lint to avoid them.

The root of my concern is a belief that typedefs can be (I might even go so far as to say they generally should be) used to capture the purpose of the function rather than the shape of the function's signature. While I understand the convenience of defining types based on the signature such as
```
typedef ZeroArgPredicate = bool Function();
```
I think it's much better to define semantically distinct types based on the usage, such as
```
typedef StartupCallback = void Function();
void startup(StartupCallback callback) { ... }

typedef ShutdownCallback = void Function();
void shutdown(ShutdownCallback callback) { ... }
```
When we need to add a parameter to `StartupCallback` it's clear which pieces of code need to be updated.

My concern is that a lint like this might suggest using `StartupCallback` in places where that's not the kind of function being defined.

However, if a function-valued variable is being passed as an argument to a parameter defined by a typedef (such as `startup` above), it would make complete sense to me that the type of the variable should be the name of the typedef and that a lint should be produced when it isn't. For example, given the declarations of `startup` and `shutdown` above, this code
```
void doSomething(void Function() f) {
}
```
would produce no lints because we don't know whether `f` should be a `StartupCallback`, `ShutdownCallback`, or neither. But this code
```
void doSomething(void Function() f) {
  startup(f);
}
```
would produce a lint because `f` is being used as a `StartupCallback` and isn't declared as such.

Not only does this remove false positives, but it also allows the lint to be very specific about which type the user should have used (and makes it easy to provide a quick fix).

Would that satisfy the use cases you have in mind, or would it miss some?

**Comment by esotericpig** [comment]
+1 as I prefer to use the typedef when defining finals, for a few reasons:

1. If the typedef changes in the future.
2. I don't want someone (or myself) adding a function parameter, when it needs to match that signature.
3. It makes it very clear what the function is for.

1 & 2 are more important for a library.

So I prefer this:

```dart
final PlaceholderWidgetBuilder placeHolder = (context, url) {
  return const CircularProgressIndicator();
};

final LoadingErrorWidgetBuilder errorWidget = (context, url, error) {
  return const FittedBox(child: Icon(Icons.error_outline));
};
```

Over having to write this:

```dart
Widget placeHolder(BuildContext context, String url) {
  return const CircularProgressIndicator();
}

Widget errorWidget(BuildContext context, String url, dynamic error) {
  return const FittedBox(child: Icon(Icons.error_outline));
}
```

The first one causes lint warning: [prefer_function_declarations_over_variables](https://dart-lang.github.io/linter/lints/prefer_function_declarations_over_variables.html).

Is there a different way to do this by using the typedef to not get this lint warning? If not, I think it'd be a good idea to change it.
